File: bench/programs/typed_buffer.py
def dot(a, b, n):
    s = 0.0
    for i in range(n): s += a[i] * b[i]
    return s
def run():
    n = 4000
    a = [i * 0.5 for i in range(n)]
    b = [i * 0.25 for i in range(n)]
    total = 0.0
    for r in range(500): total += dot(a, b, n)
    return total
print(run())

File: bench/programs/typed_buffer.go
package main

import "fmt"

//go:noinline
func dot(a, b []float64, n int) float64 {
	s := 0.0
	for i := 0; i < n; i++ {
		s += a[i] * b[i]
	}
	return s
}

func main() {
	n := 4000
	a := make([]float64, n)
	b := make([]float64, n)
	for i := 0; i < n; i++ {
		a[i] = float64(i) * 0.5
		b[i] = float64(i) * 0.25
	}
	total := 0.0
	for r := 0; r < 500; r++ {
		total += dot(a, b, n)
	}
	fmt.Println(total)
}
